src/transformers/models/big_bird/modeling_big_bird.py

```diff
@@ -0,0 +1,2629 @@
+            plan_from_length=None,
+            plan_num_rand_blocks=None,
+            output_attentions=output_attentions,
+        )
+
+        context_layer = context_layer.contiguous().view(batch_size, from_seq_length, -1)
+
+        outputs = (context_layer, attention_probs) if output_attentions else (context_layer,)
+        return outputs
+
+    @staticmethod
+    def torch_bmm_nd(inp_1, inp_2, ndim=None):
+        """ Fast nd matrix multiplication """
+        return torch.bmm(inp_1.reshape((-1,) + inp_1.shape[-2:]), inp_2.reshape((-1,) + inp_2.shape[-2:])).view(
+            inp_1.shape[: ndim - 2] + (inp_1.shape[ndim - 2], inp_2.shape[ndim - 1])
+        )
+
+    @staticmethod
+    def torch_bmm_nd_transpose(inp_1, inp_2, ndim=None):
+        """ Fast nd matrix multiplication with transpose """
+        return torch.bmm(
+            inp_1.reshape((-1,) + inp_1.shape[-2:]), inp_2.reshape((-1,) + inp_2.shape[-2:]).transpose(1, 2)
+        ).view(inp_1.shape[: ndim - 2] + (inp_1.shape[ndim - 2], inp_2.shape[ndim - 2]))
+
+    def bigbird_block_sparse_attention(
+        self,
+        query_layer,
+        key_layer,
+        value_layer,
+        band_mask,
+        from_mask,
+        to_mask,
+        from_blocked_mask,
+        to_blocked_mask,
+        num_attention_heads,
+        num_rand_blocks,
+        attention_head_size,
+        from_block_size,
+        to_block_size,
+        batch_size,
+        from_seq_length,
+        to_seq_length,
+        seed=None,
+        plan_from_length=None,
+        plan_num_rand_blocks=None,
+        output_attentions=None,
+    ):
+
+        # BigBird block-sparse attention as suggested in paper
+
+        # ITC:
+        #     global tokens: 2 x block_size
+        #     window tokens: 3 x block_size
+        #     random tokens: num_rand_tokens x block_size
+
+        # ETC:
+        #     global tokens: extra_globals_tokens + 2 x block_size
+        #     window tokens: 3 x block_size
+        #     random tokens: num_rand_tokens x block_size
+
+        # Note:
+        #     1) Currently, ETC is not supported.
+        #     2) Window size is fixed to 3 blocks & it can be changed only by
+        #     changing `block_size`.
+        #     3) Number of global blocks are fixed (2 blocks here) & global tokens can be
+        #     controlled only by `block_size`.
+
+        assert (
+            from_seq_length // from_block_size == to_seq_length // to_block_size
+        ), "Error the number of blocks needs to be same!"
+
+        # Define shorthands
+        h = num_attention_heads
+        r = num_rand_blocks
+        rsqrt_d = 1 / math.sqrt(attention_head_size)
+        b = batch_size
+        m = from_seq_length
+        n = to_seq_length
+        wm = from_block_size
+        wn = to_block_size
+
+        # generate random attention and corresponding masks
+        np.random.seed(seed)
+        if from_seq_length in [1024, 3072, 4096]:  # old plans used in paper
```

Comment:
@patrickvonplaten, I think we can further speed up by adding one extra feature of computing `randn_attn` before hand (during data preparation) instead of doing same cpu computation for every batch (since this is numpy based code completely `line 572-597`). 

But again major drawback will be that this approach won't work if someone is doing dynamic padding. For fixed seqlen there should be good increase in speed, i think.

For that, we may introduce extra function specifically for creating user's life easy.

```python
from transformers import prepare_bigbird_random_attention

random_attn = prepare_bigbird_random_attention(.....)
out = model(input_ids=..., attention_mask=..., random_attn=random_attn)
```

I can work on this one after the merge may be, if you feel it will speed up. Any view on this??

Reply:
This sounds like a great idea indeed! If it leads to a nice speed-up then I'm all for it. Let's do this after we merge :-)